ssl_test.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	return l, c, msgs, errs
+}
+
+func serverEe(t testing.TB, serverOpts SslOpts,
+	clientOpts SslOpts) (test_helpers.TarantoolInstance, error) {
+	var params string
+
+	key := serverOpts.KeyFile
+	if key != "" {
+		params += fmt.Sprintf("  \"ssl_key_file\": \"%s\"\n", key)
+	}
+
+	cert := serverOpts.CertFile
+	if cert != "" {
+		params += fmt.Sprintf("  \"ssl_cert_file\": \"%s\"\n", cert)
+	}
+
+	ca := serverOpts.CaFile
+	if ca != "" {
+		params += fmt.Sprintf("  \"ssl_ca_file\": \"%s\"\n", ca)
+	}
+
+	ciphers := serverOpts.Ciphers
+	if ciphers != "" {
+		params += fmt.Sprintf("  \"ssl_ciphers\": \"%s\"\n", ciphers)
+	}
+
+	listen := fmt.Sprintf(`"uri": "%s"
+"params":
+  "transport": "ssl"
+%s`, serverSsl, params)
+
+	return test_helpers.StartTarantool(test_helpers.StartOpts{
+		InitScript:    "config.lua",
+		Listen:        listen,
+		SslCertsDir:   "ssl_certs",
+		PingServer:    serverSsl,
+		PingTransport: "ssl",
+		PingSsl:       clientOpts,
+		WorkDir:       "work_dir_ssl",
+		User:          "test",
+		Pass:          "test",
+		WaitStart:     100 * time.Millisecond,
+		ConnectRetry:  3,
+		RetryTimeout:  500 * time.Millisecond,
+	})
+}
+
+func serverEeStop(inst test_helpers.TarantoolInstance) {
+	test_helpers.StopTarantoolWithCleanup(inst)
+}
+
+func assertConnectionLocalFail(t testing.TB, serverOpts, clientOpts SslOpts) {
+	t.Helper()
+
+	l, c, err, _, _ := createClientServerLocal(t, serverOpts, clientOpts)
+	if err == nil {
+		t.Fatalf("An unexpected connection to the server.")
+		c.Close()
+	}
+	l.Close()
+}
+
+func assertConnectionLocalOk(t testing.TB, serverOpts, clientOpts SslOpts) {
+	t.Helper()
+
+	l, c, msgs, errs := createClientServerLocalOk(t, serverOpts, clientOpts)
+	const message = "any test string"
+	c.Write([]byte(message))
+	c.Close()
+
+	recv, err := serverLocalRecv(msgs, errs)
+	l.Close()
+
+	if err != nil {
+		t.Errorf("An unexpected server error: %q", err.Error())
+	} else if recv != message {
+		t.Errorf("An unexpected server message: %q, expected %q", recv, message)
+	}
+}
+
+func assertConnectionEeFail(t testing.TB, serverOpts, clientOpts SslOpts) {
+	t.Helper()
+
+	inst, err := serverEe(t, serverOpts, clientOpts)
+	defer serverEeStop(inst)
+
+	if err == nil {
+		t.Fatalf("An unexpected connection to the server")
+	}
+}
+
+func assertConnectionEeOk(t testing.TB, serverOpts, clientOpts SslOpts) {
+	t.Helper()
+
+	inst, err := serverEe(t, serverOpts, clientOpts)
+	defer serverEeStop(inst)
+
+	if err != nil {
+		t.Fatalf("An unexpected server error %q", err.Error())
+	}
+}
+
+type test struct {
+	name       string
+	ok         bool
+	serverOpts SslOpts
+	clientOpts SslOpts
+}
+
+/*
+	Requirements from tarantool-ee manual.
+
+	For a server:
+	KeyFile - mandatory
+	CertFile - mandatory
+	CaFile - optional
+	Ciphers - optional
+
+	For a client:
+	KeyFile - optional, mandatory if server.CaFile set
+	CertFile - optional, mandatory if server.CaFile set
+	CaFile - optional,
+	Ciphers - optional
+*/
+var tests = []test{
```

Comment:
I would also add tests with Russian ciphers too (Stribog, Kuznyechik and probably other Ghost ciphers).

Reply:
I don't think so. We will test OpenSsl configuration in this case, not our code.

That is a reason why I did not test all ciphers that should work with Tarantool Enterprise. I tested only 1 cipher, which is likely to be almost everywhere.

But maybe I should add a comment about it? Something like:
```
The list of available ciphers depends on the OpenSSL configuration and is not implemented in our code.
```

Reply:
> I don't think so. We will test OpenSsl configuration in this case, not our code.

It is not so. We need a simple test to check that it is possible to use Ghost ciphers in Go connector for Tarantool. There is a high probability that our Russian customers will use Ghost ciphers with Tarantool and  I would check it.

Reply:
This is the reason why we use OpenSSL wrapper. You can add GOST ciphers to OpenSSL, but not at the application level. If we don't try to enable/disable OpenSSL's engines in our application (and we don't), then the engines configured at the system level will work.

I can configure Openssl on CI/CD like that:
https://github.com/tarantool/roadmap-internal/issues/120#issuecomment-1098897905
and test the work of GOST ciphers. But:

1. We will check just that OpenSSL can be configured according to documentation. Nothing about our library.
2. It will be hard to mantain.

I suggest not to claim that we support (or unsupport) any cipher. Because in fact we don't. All we guarantee is working with ciphers that available for OpenSSL with TLSv1.2 inside user's OS.

Reply:
if we will provide static build with OpenSSL inside, then it would need to be tested. But we don't.

Reply:
Okay, I mark it as resolved.